src/mds/MDBalancer.cc

```diff
@@ -83,14 +83,99 @@ MDBalancer::MDBalancer(MDSRank *m, Messenger *msgr, MonClient *monc) :
 {
   bal_fragment_dirs = g_conf().get_val<bool>("mds_bal_fragment_dirs");
   bal_fragment_interval = g_conf().get_val<int64_t>("mds_bal_fragment_interval");
+  last_bal_rank_mask = "0x0";
+  last_num_mdss = 0;
 }
 
 void MDBalancer::handle_conf_change(const std::set<std::string>& changed, const MDSMap& mds_map)
 {
-  if (changed.count("mds_bal_fragment_dirs"))
+  if (changed.count("mds_bal_fragment_dirs")) {
     bal_fragment_dirs = g_conf().get_val<bool>("mds_bal_fragment_dirs");
-  if (changed.count("mds_bal_fragment_interval"))
+  }
+  if (changed.count("mds_bal_fragment_interval")) {
     bal_fragment_interval = g_conf().get_val<int64_t>("mds_bal_fragment_interval");
+  }
+}
+
+void MDBalancer::handle_rank_mask_bits()
+{
+  if (mds->mdsmap->get_bal_rank_mask() == last_bal_rank_mask &&
+      mds->get_mds_map()->get_num_in_mds() == last_num_mdss) {
+    return;
+  }
+  last_bal_rank_mask = mds->mdsmap->get_bal_rank_mask();
+  last_num_mdss = mds->get_mds_map()->get_num_in_mds();
+
+  std::string bal_hex_str;
+  bal_hex_str.resize(last_bal_rank_mask.size());
+  std::transform(last_bal_rank_mask.begin(), last_bal_rank_mask.end(), bal_hex_str.begin(), ::tolower);
+
+  bool valid_hex = false;
+  num_mdss_in_rank_mask = 0;
+
+  if (bal_hex_str.substr(0, 2) == "0x") {
+    uint32_t quatet_sum = 0;
+
+    reverse(bal_hex_str.begin(), bal_hex_str.end());
+    bal_hex_str.resize(bal_hex_str.size()-2);
```

Comment:
I think this could be simplified by converting the hex string to binary `"0101..01"` string which can initialize `std::bitset`.

Reply:
Technically, the large hex string to represent MAX_MDS constituting 256bits cannot be directly converted to binary with `std::bitset`. Additonal conversion process seems somewhat inevitable.